File: #0079.word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {

        int rowSize = board.size();
        int colSize;
        for (int i = 0; i < rowSize; i++) {
            colSize = board[i].size();
            for (int j = 0; j < colSize; j++) {
                if (board[i][j] == word[0]) {
                    if (isSearchWord(board, word, 0, i, j)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

private:
    bool isSearchWord(vector<vector<char>> &board, string& word, int index, int row, int col) {

        int rowSize = board.size();
        if (row < 0 || row >= board.size()
            || col < 0 || col >= board[0].size()
            || board[row][col] != word[index]) {
            return false;
        }

        if (index + 1 == word.size()) {
            return true;
        }

        board[row][col] = '\0';
        if (isSearchWord(board, word, index+1, row+1, col)
            || isSearchWord(board, word, index+1, row-1, col)
            || isSearchWord(board, word, index+1, row, col+1)
            || isSearchWord(board, word, index+1, row, col-1)) {
            return true;
        }

        board[row][col] = word[index];
        return false;
    }
};

File: #0079.word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    const row = board.length;
    const col = board[0].length;
    const dfs = (x, y, char) => {
        if (x < 0 || y < 0
            || x >= row
            || y >= col
            || board[x][y] != word[char]) {
                return false;
        }

        if (char + 1 == word.length) {
            return true;
        }

        const cel = board[x][y];
        board[x][y] = '-';
        const results = [
            dfs(x + 1, y, char + 1),
            dfs(x - 1, y, char + 1),
            dfs(x, y + 1, char + 1),
            dfs(x, y - 1, char + 1)
        ];
        board[x][y] = cel;
        return results.some(v => v == true);
    }

    for (let i = 0; i < row; i++) {
        for (let j = 0; j < col; j++) {
            if (dfs(i, j, 0)) {
                return true;
            }
        }
    }
    return false;
};
